lib/Interface/Container.cpp

```diff
@@ -0,0 +1,287 @@
+//===- Container.cpp ------------------------------------------------------===//
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements the container descriptor.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef CORE_CONTAINER_DEF
+#define CORE_CONTAINER_DEF
+
+#include <algorithm>
+#include <cassert>
+#include <memory>
+#include <numeric>
+#include <stdexcept>
+
+#include "Interface/Container.h"
+
+/**
+ * @brief MemRef Shape Constructor. Construct a MemRef object from the data shape and initial value. The default initial value is 0.
+ * @tparam T represents the datatype to be used
+ * @tparam N represents the number of dimensions
+ */
+template <typename T, std::size_t N>
+MemRef<T, N>::MemRef(intptr_t sizes[N], T init) {
+  for (size_t i = 0; i < N; i++) {
+    this->sizes[i] = sizes[i];
+  }
+  setStrides();
+  size = product(sizes);
+  allocated = new T[size];
+  aligned = allocated;
+  std::fill(aligned, aligned + size, init);
+}
+
+
+/**
+ * @brief MemRef Array Constructor. Construct a MemRef object from the data pointer, sizes, and offset. The default offset is 0.
+ * @tparam T represents the datatype to be used
+ * @tparam N represents the number of dimensions
+ */
+template <typename T, std::size_t N>
+MemRef<T, N>::MemRef(const T *data, intptr_t sizes[N], intptr_t offset) {
+  this->offset = offset;
+  for (size_t i = 0; i < N; i++) {
+    this->sizes[i] = sizes[i];
+  }
+  setStrides();
+  size = product(sizes);
+  allocated = new T[size];
+  aligned = allocated;
+  for (size_t i = 0; i < size; i++) {
```

Comment:
Please use `std::copy` here instead of manual deep-copying 
(same advice for all other similar situations in the PR)